initial stab at replacing composite/tap-fitter

grokspawn opened the pull request:
Illustrates what removal of composite template might mean to FBC-onboarded operator-authors' process.  

This was pretty straightforward because in this case the semver template is being used over all OCP versions.  

We just replace the composite/tap-fitter segments with direct population of the catalog contribution output directories. 

For extra credit, this also uses `opm` to generate the dockerfiles so that we can eventually modify those calls (for e.g. to take advantage of binary-less catalogs in future).

grokspawn committed:
initial stab at replacing composite/tap-fitter

Signed-off-by: Jordan Keister <jordan@nimblewidget.com>

grokspawn commented on generate.sh:
```diff
@@ -62,16 +62,6 @@ then
 fi
 ln -fs "$yq_filename" yq
 
-# tap-fitter doesn't have binary downloads at the moment, so assume golang is
```
tap-fitter leverages the composite template, but earlier in the thread there was discussion of the devfile going away, so I'm not sure how much impact that actually represents.

ASzc replied:
This is ok by me. I had to remove the call to tap-fitter later on before the script was able to run, since the program is now missing

grokspawn commented on generate.sh:
```diff
@@ -119,50 +109,6 @@ EOF
     yq -e ea '.bundles | .[]' "$cfg" | xargs -n1 printf '    - image: %s@%s\n' "$bundle_reg_from"
 } > "$render_config"
 
-echo "-> Defining supported OCP versions" >&2
```
This is pulling definition of the catalogs/contributions schemas for the composite templates.  Instead, we just uniformly apply the semver expansion to all OCP_VER.

ASzc replied:
Makes sense

grokspawn commented on generate.sh:
```diff
@@ -207,60 +154,60 @@ then
 for ocp_ver in "${ocp_versions[@]}"
 do
-    # Soon this might be able to be simplified, as "binaryless" container, FROM scratch
+    pushd catalog/$ocp_ver
```
I took some liberty here to show how to use opm to generate the dockerfile for you, even with the image name variability that we were saddled with. 😞

ASzc replied:
I'm ok with opm generate being used here, although there are some differences in the output that make me uneasy

```diff
diff --git a/catalog/v4.11/Dockerfile b/catalog/v4.11/Dockerfile
index 73e0087..dee47c1 100644
--- a/catalog/v4.11/Dockerfile
+++ b/catalog/v4.11/Dockerfile
@@ -7,7 +7,7 @@ ENTRYPOINT ["/bin/opm"]
 CMD ["serve", "/configs", "--cache-dir=/tmp/cache"]
 
 # Copy declarative config root into image at /configs and pre-populate serve cache
-ADD rhbk-operator /configs/rhbk-operator
+ADD rhbk-operator /configs
 RUN ["/bin/opm", "serve", "/configs", "--cache-dir=/tmp/cache", "--cache-only"]
 
 # Set DC-specific label for the location of the DC root directory
```

grokspawn replied:
This results in the generated FBC from (for e.g. `./catalog/v4.13/rhbk-operator/catalog.yaml`) getting set in the image as /configs/catalog.yaml. 
For a single-package contribution this is fine, and only differs by omitting the intermediate $operator_name directory. 
If that's important to you, then we just have to move some elements around.

ASzc replied:
As long as IIB and opm are ok with this, I don't particularly care. I never inspect the index image directly, only through `opm render`

grokspawn replied:
Understood.  opm considers the specific filesystem hierarchy as FBC and aggregates it, so location doesn't matter to opm... just whether the FBC follows the schema.

grokspawn commented on generate.sh:
```diff
@@ -207,60 +154,60 @@ then
-# Set DC-specific label for the location of the DC root directory
-# in the image
-LABEL operators.operatorframework.io.index.configs.v1=/configs
+    cat <<EOF
```
Here's the replacement devfile generation, instead of using the composite template.

ASzc replied:
Even more so than the Dockerfile, there are some differences in the output now. The `kubernetes:` line seems like a syntax error now? Or is that part of intended changes?

```diff
diff --git a/catalog/v4.11/devfile.yaml b/catalog/v4.11/devfile.yaml
index b7043d4..6897155 100644
--- a/catalog/v4.11/devfile.yaml
+++ b/catalog/v4.11/devfile.yaml
@@ -1,7 +1,7 @@
 schemaVersion: 2.2.0
 metadata:
-  name: fbc-v4.11
-  displayName: FBC v4.11
+  name: fbc-rhbk-operator
+  displayName: FBC rhbk-operator
   description: 'File based catalog'
   language: fbc
   provider: Red Hat
@@ -10,17 +10,17 @@ components:
     image:
       imageName: ""
       dockerfile:
-        uri: Dockerfile
+        uri: .Dockerfile
         buildContext: ""
-  - name: kubernetes
-    kubernetes:
-      inlined: placeholder
+  - kubernetes:
+    inlined: "placeholder"
     attributes:
-      deployment/container-port: 50051
-      deployment/cpuRequest: 100m
-      deployment/memoryRequest: 512Mi
-      deployment/replicas: 1
-      deployment/storageRequest: "0"
+    - deployment:
+        container-port: 50051
+        cpuRequest: "100m"
+        memoryRequest: "512Mi"
+        replicas: 1
+        storageRequest: 0
 commands:
   - id: build-image
     apply:
```

grokspawn replied:
Dropping the devfile stuff, since evidently it's no longer needed.

ASzc reviewed (commented):
I tried it with rhbk-fbc repo, and it seemed to work with a few changes. However since the output isn't exactly the same, and my experience with Konflux is that it's very touchy, I'm going to need some explanation for the output differences

grokspawn commented:
I hadn't actually run the initial changes before.

I made some additional updates
1. made the script mac-friendly :) 
2. fixed the dockerfiles output paths
3. created directory hierarchy that was previously created by composite-template/tap-fitter
4. fixed a few `rhbk-operator` literals

I noticed the script depends on the presence of the replacements in the config file, and I wasn't sure that was intentional since the docs mention conditional dependency.  And it sounded like a catalog use dependency rather than a generate.sh dependency.

ASzc commented:
Thanks for the updates! I ran it against rhbk-fbc and it seems to be good output. There's one difference in the Dockerfile still, but we've discussed above that this isn't an issue.

Regarding `replacements` being required in the config: I suppose it is in reality, as part of the assumption that users of fbc-utils have a RHBK-like use case where there are internal-only images are commonly referenced by the FBC. This could be reworked but is probably too much to try to roll into this PR.

With a pure Konflux build (i.e. NOT what RHBK is doing at present), all the images end up in quay.io? Or do they have some private storage that would require `replacements` too?

grokspawn commented:
+1 to trying to work on replacements in a different PR.  That would IMO distract from the functional changes here.

>With a pure Konflux build (i.e. NOT what RHBK is doing at present), all the images end up in quay.io? Or do they have some private storage that would require replacements too?

I don't know what the convention is, but it might depend on the catalog.  The official ones end up at registry.redhat.io, but I don't know where the fragment-minicats live.

ASzc commented:
Merged, thanks again!

ASzc commented:
I filed #3 to track the replacements issue
